### PR TITLE
modules101 article

### PR DESCRIPTION
**Pull request by dmitrypol**
### Description

Created new blog article on building modules in C or Rust
 
### Issues Resolved

Improved instructions in README

### Check List
- [x] Commits are signed per the DCO using `--signoff`

By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

### CODE REVIEW

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:56]: Ideally we would use the valkeymodule-rs SDK correctly in this blog post, which is to take a dependency on it and then build it. We don't really want people using the sdk this way, and in the long term people might get confused by having this legacy post.

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:78]: It would probably be useful to explain what this is doing, not just say that "the syntax is different".

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [reply _posts/2024-04-26-modules-101.markdown:56]: Can we also cleanup the references to redis before publishing this? I think it would be better for branding to limit the references to Redis in this post.

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:10]: I think for branding purposes, we should omit Redis completely from this post, unless specifically talking about the license changes and why the project exists.

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:7]: What are our categories?

Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:7]: good question.  I updated the category to `modules` but I can omit it.  It becomes part of the URL so we don't want to change it afterwards or it will direct links

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:10]: Completely agree on branding.  I removed all references except the ones required by SDK code.  Working to update SDK but that's going to take a bit of time.

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:56]: Agree on branding.  I am working to get SDK into crates.io but I want to cleanup before publishing it as crate.

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:78]: added details

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown]: @madolson - per our chat I published https://crates.io/crates/valkey-module and updated this blog.

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown]: updated blog to use Valkey* vars.  You can see SDK changes in https://github.com/valkey-io/valkeymodule-rs/pull/8

File: _authors/dmitrypol.markdown
Comment (stockholmux) [inline _authors/dmitrypol.markdown:10]: He is very excited to be part of **the** Valkey project.

File: _posts/2024-04-26-modules-101.markdown
Comment (stockholmux) [inline _posts/2024-04-26-modules-101.markdown:12]: > This articles expects the audience to be

article (singular)

File: _posts/2024-04-26-modules-101.markdown
Comment (stockholmux) [inline _posts/2024-04-26-modules-101.markdown:38]: initalizing -> initializing

File: _posts/2024-04-26-modules-101.markdown
Comment (stockholmux) [inline _posts/2024-04-26-modules-101.markdown:49]: put the patterns (*.so/*.xo) in backticks

File: _posts/2024-04-26-modules-101.markdown
Comment (stockholmux) [inline _posts/2024-04-26-modules-101.markdown:54]: To install **the** valkey-module SDK run cargo add valkey-module. It will create or update **the** Cargo.lock file.

File: _posts/2024-04-26-modules-101.markdown
Comment (stockholmux) [inline _posts/2024-04-26-modules-101.markdown:86]: "Since we not using ctx or args..." -> "Since we're not using ctx or args..."

File: _posts/2024-04-26-modules-101.markdown
Comment (stockholmux) [inline _posts/2024-04-26-modules-101.markdown:93]: Go back into **the** Valkey repo folder and run make to compile the Valkey code.  Then add these lines to the bottom of  **the** `valkey.conf` file.

**Review by stockholmux — CHANGES_REQUESTED** [review]
A few grammar things.

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:12]: udpated

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:38]: udpated

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:49]: done

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:54]: updated

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:86]: updated

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:93]: updated

File: _authors/dmitrypol.markdown
Comment (madolson) [inline _authors/dmitrypol.markdown:8]:
```suggestion
Dmitry is a Consulting Member of Technical Staff at Oracle Cloud.  
```

Unless you are the only one?

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:12]:
```suggestion
The idea of modules is to allow adding extra features (such as new commands and data types) to Valkey without making changes to the core code.  Modules are a special type of code distribution called a shared library, which can be loaded by other programs at runtime and executed.  Modules can be written in C or other languages that have C bindings.  In this article we will go through the process of building simple modules in C and Rust (using the Valkey Module Rust SDK).  This article expects the audience to be at least somewhat familiar with C, Rust and Valkey.  
```

@dmitrypol @stockholmux I would also like to suggest we enforce one sentence per line, it becomes much easier to review. 

The blog also assumes knowledge of git and cloning. I would suggest either adding more detailed structure for cloning Valkey so that it keeps the bar low.

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:16]:
```suggestion
If we clone the [Valkey](https://github.com/valkey-io/valkey) repo we will find numerous examples in `src/modules`.  Let's create a new file `module1.c` in the same folder.  
```

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:38]: I suppose there is a question of what is 101 vs what isn't. "All we are doing is initializing module1" is doing a lot of heavy lifting in this sentence. What is the initialization code? What is creating the command? Do we want to explain that here or maybe in a follow up.

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:49]:
```suggestion
Run `make module1.so` inside `src/modules` folder.  This will compile our module in the src/modules folder. 
```

No need to generate everything.

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:56]:
```suggestion
Modify `Cargo.toml` to specify the crate-type to be "cdylib", which will tell cargo to build the target as a shared library .  Read [Rust docs](https://doc.rust-lang.org/reference/linkage.html) to understand more about `crate-type`.
```

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:93]:
```suggestion
Go back into the Valkey repo folder and run `make` to compile the Valkey code.  Then add these lines to the bottom of the `valkey.conf` file.
```

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:127]:
```suggestion
We can now run both modules side by side and if we modify either C or RS file, recompile the code and restart `valkey-server` we will get the new functionality.  Alternatively to specifying modules in .conf file we can use MODULE LOAD and UNLOAD from valkey-cli at runtime.  
```

I don't like blogs that say, "Alternatively do X/Y" without explaining it. So would prefer to either drop the reference or show an example of it.

**Review by madolson — COMMENTED** [review]
Did a path for wording, doing a second path to follow it as a user now.

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:69]: We are missing the Valkey dependency

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:54]: `cargo add valkey-module` is not a comand.

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:69]: Right now I get an issue building this, 

```
error[E0432]: unresolved import `valkey_module::ValkeyValue`
 --> src/lib.rs:4:58
  |
4 | use valkey_module::{Context, ValkeyResult, ValkeyString, ValkeyValue};
  |                                                          ^^^^^^^^^^^ no `ValkeyValue` in the root

For more information about this error, try `rustc --explain E0432`.
```

Seems like Valkeyvalue is still called a RedisModule in the published version.

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:97]: Should probably clarify the "path_to_*" needs to be updated.

File: _authors/dmitrypol.markdown
Comment (dmitrypol) [reply _authors/dmitrypol.markdown:8]: thank you

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:12]: updated

Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:12]: not sure what you mean by one sentence per line.  That will impact how the text is presented in the published blog.

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:16]: done

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:38]: added more details.

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:49]: fair point, updated

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:56]: updated

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:93]: updated

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:69]: This needs valkey-module 0.1.0 crate which I will publish to crates.io once you approve https://github.com/valkey-io/valkeymodule-rs/pull/8.  I can publish it now as is but then if we make changes to the SDK we will need to bump version.

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:97]: updated

File: _authors/dmitrypol.markdown
Comment (stockholmux) [reply _authors/dmitrypol.markdown]: I thought you were the singular important one ;)

File: _posts/2024-04-26-modules-101.markdown
Comment (stockholmux) [reply _posts/2024-04-26-modules-101.markdown]: @dmitrypol Create a new line for each sentence (not in a code block). Markdown create paragraph breaks after two new lines, so it'll look fine in the resulting blog post.

Example:
```
This is my first sentence. 
This is my second sentence.

This is a new paragraph.
```

Will render like:
> This is my first sentence. This is my second sentence.
> 
> This is a new paragraph.

@madolson I _really_ need to setup a style guide and perhaps a action to check for these. I'll add it to the ever growing to-do list.

Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown]: I thought it does it too but when I make the change locally and run `bundle exec jekyll serve` it generates the break after one line. I will make the change and hopefully it will look fine

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [reply _posts/2024-04-26-modules-101.markdown]: Yeah, I updated cargo and it worked. I apparently installed cargo in 2020

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:150]:
```suggestion
As an alternative to specifying modules in `valkey.conf` file, we can use `MODULE LOAD` and `UNLOAD` from `valkey-cli` to update the server.
```

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:101]:
```suggestion
We are using the external crate `valkey_module` with [Rust macros](https://doc.rust-lang.org/book/ch19-06-macros.html) and passing it variables like `name` and `version`.  
```

File: _posts/2024-04-26-modules-101.markdown
Comment (madolson) [inline _posts/2024-04-26-modules-101.markdown:107]:
```suggestion
The build will produce *.so files on Linux and *.dll files on Windows.  
```

**Review by madolson — APPROVED** [review]
Just nitpicks, so approving if you want to update and merge.

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:150]: updated

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:101]: updated

File: _posts/2024-04-26-modules-101.markdown
Comment (dmitrypol) [reply _posts/2024-04-26-modules-101.markdown:107]: updated

**Review by madolson — APPROVED** [review]

**Review by stockholmux — APPROVED** [review]
I'm going to approve and merge. 

**Next step:** Double check for publishing suitability, make changes needed for publishing (probably just change publishing date.)

**Pull request merged**
